Azure/azure-cli
# [Role] az role assignment create/update: Support `--description`, `--condition` and `--condition-version`

Close https://github.com/Azure/azure-cli/issues/14552

## `az role assignment create`

### Design

New arguments to add to the existing arguments:

```
Arguments
    --description             : Description of the role assignment.
    --condition               : Condition under which the user can be granted permission
    --condition-version       : Version of the condition syntax
```

### Examples

```sh
# line breaks for legibility
az role assignment create --assignee admin4@AzureSDKTeam.onmicrosoft.com
                          --role reader
                          --scope /subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1
                          --description "Role assignment foo to check on bar"
                          --condition "@Resource[Microsoft.Storage/storageAccounts/blobServices/containers:Name] stringEquals 'foo'"
                          --condition-version "2.0"
```

Sample Output:

```json
{
  "canDelegate": null,
  "condition": "@Resource[Microsoft.Storage/storageAccounts/blobServices/containers:Name] stringEquals 'foo'",
  "conditionVersion": "2.0",
  "description": "Role assignment foo to check on bar",
  "id": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1/providers/Microsoft.Authorization/roleAssignments/d89f022c-f12f-4fb5-9d90-afb9c1f4fd83",
  "name": "d89f022c-f12f-4fb5-9d90-afb9c1f4fd83",
  "principalId": "182c8534-f413-487c-91a3-7addc80e35d5",
  "principalType": "User",
  "resourceGroup": "rg1",
  "roleDefinitionId": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleDefinitions/acdd72a7-3385-48ef-bd42-f606fba81ae7",
  "scope": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1",
  "type": "Microsoft.Authorization/roleAssignments"
}
```

### Checks

#### Description

* `description` is a string.
* If `description` is empty, replace with null.

#### Condition

* `condition` is a string.
* If `condition` is empty replace with null.

#### Condition-Version

* `condition-version` is a string.
* `condition-version` must be > `1.0`.
  - The service already has this check: `The specified role assignment ConditionVersion '0.1' is not supported`.
* If `condition-version` is empty and `condition` is set, `condition-version` defaults to `2.0`.
* If `condition-version` is set, `condition` must be set as well.


## `az role assignment update`

### Design

``` 
Command
    az role assignment update : update an existing role assignment for a user, group, or service
    principal.

Arguments
    --role-assignment   [Required] :  Description of a role as JSON, or a path to a file containing a JSON description.
```

### Example

```sh
az role assignment update --role-assignment '{
  "canDelegate": null,
  "condition": "@Resource[Microsoft.Storage/storageAccounts/blobServices/containers:Name] stringEquals '"'"'foo'"'"'",
  "conditionVersion": "2.0",
  "description": "Role assignment foo to check on bar",
  "id": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1/providers/Microsoft.Authorization/roleAssignments/d89f022c-f12f-4fb5-9d90-afb9c1f4fd83",
  "name": "d89f022c-f12f-4fb5-9d90-afb9c1f4fd83",
  "principalId": "182c8534-f413-487c-91a3-7addc80e35d5",
  "principalType": "User",
  "resourceGroup": "rg1",
  "roleDefinitionId": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleDefinitions/acdd72a7-3385-48ef-bd42-f606fba81ae7",
  "scope": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1",
  "type": "Microsoft.Authorization/roleAssignments"
}'

# or
az role assignment update --role-assignment "C:\Path\to\JSON\roleAssignment.json
```

Sample Output:

```json
{
  "canDelegate": null,
  "condition": "@Resource[Microsoft.Storage/storageAccounts/blobServices/containers:Name] stringEquals 'foo'",
  "conditionVersion": "2.0",
  "description": "Role assignment foo to check on bar",
  "id": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1/providers/Microsoft.Authorization/roleAssignments/d89f022c-f12f-4fb5-9d90-afb9c1f4fd83",
  "name": "d89f022c-f12f-4fb5-9d90-afb9c1f4fd83",
  "principalId": "182c8534-f413-487c-91a3-7addc80e35d5",
  "principalType": "User",
  "resourceGroup": "rg1",
  "roleDefinitionId": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleDefinitions/acdd72a7-3385-48ef-bd42-f606fba81ae7",
  "scope": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1",
  "type": "Microsoft.Authorization/roleAssignments"
}
```

### Checks

Same as for `role assignment create` with some small exceptions/additions:

* Only `description`, `condition` and `conditionVersion` can be modified.
  - This requires lots of efforts for diffing the original REST response and user input and should be implemented on the service side. NOT IMPLEMENTED in CLI yet.
* if no role assignment exists with the indicated properties, throw an exception indicating as such.
* `conditionVersion` cannot be downgraded to `1.X`.

## Testing Guide

```sh
az group create -l westus -n rg1

az role assignment create --assignee admin4@AzureSDKTeam.onmicrosoft.com --role reader --scope /subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1 --description "Role assignment foo to check on bar" --condition "@Resource[Microsoft.Storage/storageAccounts/blobServices/containers:Name] stringEquals 'foo'" --condition-version "2.0"

az role assignment update --role-assignment '{
  "canDelegate": null,
  "condition": "@Resource[Microsoft.Storage/storageAccounts/blobServices/containers:Name] stringEquals '"'"'foo'"'"'",
  "conditionVersion": "2.0",
  "description": "Role assignment foo to check on bar",
  "id": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1/providers/Microsoft.Authorization/roleAssignments/9e8947f9-a813-4003-bbdd-98fa57d4dca0",
  "name": "9e8947f9-a813-4003-bbdd-98fa57d4dca0",
  "principalId": "182c8534-f413-487c-91a3-7addc80e35d5",
  "principalName": "admin4@AzureSDKTeam.onmicrosoft.com",
  "principalType": "User",
  "resourceGroup": "rg1",
  "roleDefinitionId": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleDefinitions/acdd72a7-3385-48ef-bd42-f606fba81ae7",
  "roleDefinitionName": "Reader",
  "scope": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1",
  "type": "Microsoft.Authorization/roleAssignments"
}'

az role assignment delete -g rg1

az role assignment create --assignee admin4@AzureSDKTeam.onmicrosoft.com --role reader --scope /subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/rg1 --description "Role assignment foo to check on bar" --condition "@Resource[Microsoft.Storage/storageAccounts/blobServices/containers:Name] stringEquals 'foo'" --condition-version "2.0" > output.json

az role assignment update --role-assignment output.json

az group delete -n rg1
```
